tracer/src/Datadog.Trace/Debugger/Symbols/SymbolBatchUploadApi.cs

```diff
@@ -0,0 +1,121 @@
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+#nullable enable
+using System;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Datadog.Trace.Agent;
+using Datadog.Trace.Agent.DiscoveryService;
+using Datadog.Trace.Agent.Transports;
+using Datadog.Trace.Debugger.Sink;
+using Datadog.Trace.Logging;
+
+namespace Datadog.Trace.Debugger.Symbols
+{
+    internal class SymbolBatchUploadApi : IBatchUploadApi
+    {
+        private static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor<SymbolBatchUploadApi>();
+        private readonly IApiRequestFactory _apiRequestFactory;
+        private readonly ArraySegment<byte> _eventMetadata;
+        private string? _endpoint;
+        private IDiscoveryService? _discoveryService;
+
+        private SymbolBatchUploadApi(IApiRequestFactory apiRequestFactory, IDiscoveryService discoveryService, ArraySegment<byte> eventMetadataMetadata)
+        {
+            _apiRequestFactory = apiRequestFactory;
+            _eventMetadata = eventMetadataMetadata;
+            _discoveryService = discoveryService;
+            _discoveryService.SubscribeToChanges(ConfigurationChanged);
+        }
+
+        private static ArraySegment<byte> GetEventAsArraySegment(string? serviceName)
+        {
+            var sb = new StringBuilder();
+            sb.Append(@"{");
+            sb.Append(@"""ddsource"": ""dd_debugger"",");
+            sb.Append(@$"""service"": ""{serviceName}"",");
+            sb.Append(@$"""runtimeId"": ""{Tracer.RuntimeId}""");
+            sb.Append(@"}");
+            var eventMetadata = sb.ToString();
+
+            var count = Encoding.UTF8.GetByteCount(eventMetadata);
+            var eventAsBytes = new byte[count];
+            Encoding.UTF8.GetBytes(eventMetadata, 0, eventMetadata.Length, eventAsBytes, 0);
+            return new ArraySegment<byte>(eventAsBytes);
+        }
+
+        public static IBatchUploadApi Create(IApiRequestFactory apiRequestFactory, IDiscoveryService discoveryService, string? serviceName)
+        {
+            try
+            {
+                var eventMetadata = GetEventAsArraySegment(serviceName);
+                return new SymbolBatchUploadApi(apiRequestFactory, discoveryService, eventMetadata);
+            }
+            catch (Exception e)
+            {
+                Log.Error(e, "Failed to create {Class}. Creating instead {NoOpClass}.", nameof(SymbolBatchUploadApi), nameof(NoOpSymbolBatchUploadApi));
+                return new NoOpSymbolBatchUploadApi();
+            }
+        }
+
+        private void ConfigurationChanged(AgentConfiguration config)
+        {
+            _endpoint = config.SymbolDbEndpoint;
+            if (!string.IsNullOrEmpty(Volatile.Read(ref _endpoint)))
+            {
+                _discoveryService!.RemoveSubscription(ConfigurationChanged);
```

Comment:
This seems to imply that the user may only upgrade `datadog-agent`, never downgrade - but they might? It's probably not the worst thing in the world, since IIUC we'll just attempt to send a symbol upload to an endpoint that doesn't exist in this case, but just thought I'd call it out.

Reply:
This is an important point. Do you think we should support such a case? If so, as it currently works we can avoiding sending in case of a downgrade **but still do the work of getting all the symbols**. More significant work is needed to completely prevent the operation of the symbol extractor.

Reply:
I think we're fine to just ignore this tbh, so I'll go ahead and resolve this. Just wanted to call it out - in the rare case that this happens, we'll get an error in the logs, and it shouldn't cause any particular problems.